File: packages/core/src/routing/state.ts
/**
 * Optional persistence for learned routing state (health, outcomes, circuit
 * breakers). Implement over Redis/Postgres/anything; inject via
 * `EngineOptions.stateStore`. Without it, all learning stays in-process and
 * dies with the engine instance.
 *
 * Semantics are deliberately simple: namespaced JSON snapshots written
 * behind-the-scenes after mutations (debounced), loaded once at startup.
 * Last-writer-wins per snapshot — replicas converge on fresh-enough data,
 * not linearizable truth. Failures to read/write never break routing.
 */

export interface RouterStateStore {
  get(key: string): Promise<string | null | undefined> | string | null | undefined;
  set(key: string, value: string, ttlMs: number): Promise<void> | void;
}

/** Default key under which the engine persists its combined snapshot. */
export const DEFAULT_STATE_KEY = "ai-router:state:v1";

/**
 * Cheap request-shape classifier for workload-aware routing. Produces the
 * `task` bucket used by quality-first ordering when the caller does not
 * name one (`routing.task`). Heuristic, deterministic, no ML: buckets map
 * to routing-relevant workload shapes, not semantics.
 *
 * `estimatedInputTokens` comes from the caller (engine passes its own
 * `estimateTokens` result) to keep this module dependency-free.
 */
import type { ChatRequest } from "../types.js";

export function inferTask(req: ChatRequest, estimatedInputTokens?: number): string {
  if (req.tools && req.tools.length > 0) return "tool-use";
  for (const msg of req.messages) {
    if (Array.isArray(msg.content) && msg.content.some((p) => p.type === "image_url")) {
      return "vision";
    }
  }
  if (req.response_format?.type === "json_schema" || req.response_format?.type === "json_object") {
    return "structured";
  }
  if (req.reasoning_effort !== undefined) return "reasoning";
  const tokens = estimatedInputTokens ?? roughTokens(req);
  if (tokens > 8_000) return "long-context";
  return "chat";
}

/** Fallback char heuristic (~3.5 chars/token) matching engine.estimateTokens closely enough. */
function roughTokens(req: ChatRequest): number {
  let chars = 0;
  for (const msg of req.messages) {
    if (typeof msg.content === "string") chars += msg.content.length;
    else if (Array.isArray(msg.content)) chars += msg.content.length;
  }
  return Math.ceil(chars / 3.5) + 1;
}

File: packages/python/src/ai_router/routing/state.py
"""Router state store interface and task inference."""

from __future__ import annotations

import math
from typing import Any, Protocol, runtime_checkable

DEFAULT_STATE_KEY = "ai-router:state:v1"


@runtime_checkable
class RouterStateStore(Protocol):
    async def get(self, key: str) -> str | None: ...

    async def set(self, key: str, value: str, ttl_ms: int) -> None: ...


def _rough_tokens(req: Any) -> int:
    chars = 0
    messages = getattr(req, "messages", None) if not isinstance(req, dict) else req.get("messages")
    if messages:
        for msg in messages:
            content = getattr(msg, "content", None) if not isinstance(msg, dict) else msg.get("content")
            if isinstance(content, str):
                chars += len(content)
            elif isinstance(content, list):
                chars += sum(len(p.get("text", "") if isinstance(p, dict) else getattr(p, "text", "")) for p in content)
    return math.ceil(chars / 3.5) + 1


def infer_task(req: Any, estimated_input_tokens: int | None = None) -> str:
    """Classifies a request into a workload task bucket."""
    tools = getattr(req, "tools", None) if not isinstance(req, dict) else req.get("tools")
    if tools and len(tools) > 0:
        return "tool-use"

    messages = getattr(req, "messages", None) if not isinstance(req, dict) else req.get("messages")
    if messages:
        for msg in messages:
            content = getattr(msg, "content", None) if not isinstance(msg, dict) else msg.get("content")
            if isinstance(content, list):
                for p in content:
                    p_type = getattr(p, "type", None) if not isinstance(p, dict) else p.get("type")
                    if p_type == "image_url":
                        return "vision"

    rf = getattr(req, "response_format", None) if not isinstance(req, dict) else req.get("response_format")
    if isinstance(rf, dict):
        if rf.get("type") in ("json_schema", "json_object"):
            return "structured"
    elif rf == "json_object":
        return "structured"

    reasoning_effort = (
        getattr(req, "reasoning_effort", None) if not isinstance(req, dict) else req.get("reasoning_effort")
    )
    if reasoning_effort is not None:
        return "reasoning"

    tokens = estimated_input_tokens if estimated_input_tokens is not None else _rough_tokens(req)
    if tokens > 8000:
        return "long-context"

    return "chat"
